plugins/catalog-backend/src/ingestion/processors/util/github.ts

```diff
@@ -0,0 +1,285 @@
+ * @param org The slug of the org to read
+ */
+export async function getOrganizationUsers(
+  client: typeof graphql,
+  org: string,
+): Promise<{ users: UserEntity[] }> {
+  const users: UserEntity[] = [];
+
+  // There is no user -> teams edge, so we leave the memberships empty for
+  // now and let the team iteration handle it instead
+  let cursor: string | undefined = undefined;
+  const query = `
+      query users($org: String!, $cursor: String) {
+        organization(login: $org) {
+          membersWithRole(first: 100, after: $cursor) {
+            pageInfo { hasNextPage, endCursor }
+            nodes { avatarUrl, bio, email, login, name }
+          }
+        }
+      }`;
+
+  for (let i = 0; i < 100 /* just for sanity */; ++i) {
+    const response: QueryResponse = await client(query, {
+      org,
+      cursor,
+    });
+
+    const connection = response.organization?.membersWithRole;
+    if (!connection) {
+      throw new Error(`Found no organization named ${org}`);
+    }
+
+    for (const user of connection.nodes) {
+      const entity: UserEntity = {
+        apiVersion: 'backstage.io/v1alpha1',
+        kind: 'User',
+        metadata: {
+          name: user.login,
+          annotations: {
+            'github.com/user-login': user.login,
+          },
+        },
+        spec: {
+          profile: {},
+          memberOf: [],
+        },
+      };
+
+      if (user.bio) entity.metadata.description = user.bio;
+      if (user.name) entity.spec.profile!.displayName = user.name;
+      if (user.email) entity.spec.profile!.email = user.email;
+      if (user.avatarUrl) entity.spec.profile!.picture = user.avatarUrl;
+
+      users.push(entity);
+    }
+
+    const { hasNextPage, endCursor } = connection.pageInfo;
+    if (!hasNextPage) {
+      break;
+    } else {
+      cursor = endCursor;
+    }
+  }
+
+  return { users };
+}
+
+/**
+ * Gets all the teams out of a GitHub organization.
+ *
+ * Note that the teams will not have any relations apart from parent filled in.
+ *
+ * @param client An octokit graphql client
+ * @param org The slug of the org to read
+ */
+export async function getOrganizationTeams(
+  client: typeof graphql,
+  org: string,
+): Promise<{
+  groups: GroupEntity[];
+  groupMemberUsers: Map<string, string[]>;
+}> {
+  const groups: GroupEntity[] = [];
+  const groupMemberUsers = new Map<string, string[]>();
+
+  let cursor: string | undefined = undefined;
+  const query = `
+    query teams($org: String!, $cursor: String) {
+      organization(login: $org) {
+        teams(first: 100, after: $cursor) {
+          pageInfo { hasNextPage, endCursor }
+          nodes {
+            slug
+            combinedSlug
+            parentTeam { slug }
+            members(first: 100, membership: IMMEDIATE) {
+              pageInfo { hasNextPage, endCursor }
+              nodes { login }
+            }
+          }
+        }
+      }
+    }`;
+
+  for (let i = 0; i < 100 /* just for sanity */; ++i) {
```

Comment:
I bet paging could be handled with a higher order function. I actually think I may have something kicking around for this particular case. Will try and dig it up.

Reply:
This is something I wiped together a few weeks back essentially doing the same thing against the GitHub API. I take no disclaimer towards this though 🙂 

```ts
type PaginatedQueryArgs<T> = {
  query: any;
  mapPage: (data: any) => { data: T; next: string };
  pageSize?: number;
  variables?: any;
};

export async function queryAllPages<T>({
  query,
  mapPage,
  pageSize = 50,
  variables,
}: PaginatedQueryArgs<T[]>) {
  const first = pageSize;
  let after: string = null;
  let all: T[] = [];

  // eslint-disable-next-line no-constant-condition
  while (true) {
    const page = await queryPage({ query, mapPage, first, after, variables });
    all = [...all, ...page.data];
    after = page.next;

    if (!after) {
      break;
    }
  }

  return all;
}

type PageQueryArgs<T> = {
  query: any;
  mapPage: (data: any) => { data: T[]; next: string };
  first;
  after;
  variables?: any;
};

export async function queryPage<T>({
  query,
  mapPage,
  first = 10,
  after = null,
  variables,
}: PageQueryArgs<T>) {
  logger.log(`fetching first: ${first}, after: ${after}`);

  const client = createClient();

  return client
    .query({ query, variables: { ...variables, first, after } })
    .then(r => r.data)
    .then(mapPage);
}
```

NOTE: The client is actually an apollo client in this case, but I suspect it could be adapted to `octokit`.

Reply:
Yep I have the same thing written here actually :)

Reply:
Similar 😄. The idea I had was this could work for any paged query and also avoid having a magic number for the max number of pages.

Reply:
No I meant, i had the same very similar higher order function implemented locally but not committed yet - now it's in though :) pushed a couple minutes ago